### DISCUSSION TITLE
How to execute WSL commands from a windows batch file?

### DISCUSSION THREAD

**Discussion by tinmac** [microsoft/WSL#6128] [Q&A] [answered]
In Visual Studio 2019 im running an 'AfterPublish' event which calls a windows batch file which uses SCP to copy the published files to the Linux server.

SCP works fine from the batch file but id like to use rsync, which has options to only copy modified files, cutting down time and bandwidth.

RSYNC however is part of WSL, I can call it from a cmd prompt fine.

 But how do I call 'wsl rsync...'  from within a windows batch file?

**Comment by craigloewen-msft** [comment]
You can call it just by using `wsl.exe <yourCommand here>`. For example adding the line: `wsl.exe echo "Hello World"` will print Hello World, and running `wsl.exe -u root apt update` will run the apt binary and update your packages as the root user in your distro. 

Does this help answer your question?

**Comment by tinmac** [comment, marked as answer]
Hello @craigloewen-msft 

Thanks for the reply but no thats not working for me, when I call a script (eg test.bat or test.ps1) as a postbuild event in VS Properties -> Build Events eg

`"C:\Windows\System32\WindowsPowerShell\v1.0\Powershell.exe" -file .\test.ps1`

In the VS Output window I get 'The term 'wsl' is not recognized as the name of a cmdlet,...'

If I run the same test.bat or test.ps1 form cmd or powershell it works fine.

Im on WSL Ubuntu 20.04, Windows 2004 - 19041.572

Any pointers?

**Reply by craigloewen-msft** [reply]
Ah this is because Visual Studio is running as a 32 bit app, and WSL in System32 is a 64 bit app. 

I was able to get it running using this command under post build events in Visual Studio:
`powershell.exe /c C:\Windows\sysnative\wsl.exe echo Hello World`

Let me know if that fixes it for you!

![image](https://user-images.githubusercontent.com/42221804/97617720-5b95ac80-19db-11eb-8332-daf5ed3c554f.png)

![image](https://user-images.githubusercontent.com/42221804/97617742-63555100-19db-11eb-96ad-aed293343105.png)

**Reply by tinmac** [reply]
Ah right, I see, thanks for the clarification. 

I found [this ](https://www.samlogic.net/articles/sysnative-folder-64-bit-windows.htm)article which also helped to explain the 32/64 bit access issue and cement the concept. 

Many thanks for your help @craigloewen-msft

**Comment by Laureanojb** [comment]
what about this (the format worked out perfectly to me):
bash -c "cd /mnt/c/projects/Test/jekyll/help; bundle exec jekyll serve;"
From here [https://weblog.west-wind.com/posts/2020/May/19/Using-WSL-to-Launch-Several-Bash-Command-from-an-Application](https://weblog.west-wind.com/posts/2020/May/19/Using-WSL-to-Launch-Several-Bash-Command-from-an-Application)
I needed to do a workaround with some batch files to open 6 processes from start up...

**Reply by Laureanojb** [reply]
Many thanks to Rick Strahl !!
